cpp/tests/experimental/find/find_and_combine_segments_test.cu

```diff
@@ -0,0 +1,361 @@
+{
+  using T       = TypeParam;
+  using index_t = std::size_t;
+  using P       = vec_2d<T>;
+  using S       = segment<T>;
+
+  auto segment_offsets = make_device_vector<index_t>({0, 2});
+  auto segments =
+    make_device_vector<segment<T>>({S{P{0.0, 0.0}, P{0.0, 0.5}}, S{P{0.0, 0.5}, P{0.0, 1.0}}});
+  auto merged_flag = rmm::device_vector<uint8_t>(segments.size());
+
+  auto expected_segments =
+    make_device_vector<segment<T>>({S{P{0.0, 0.0}, P{0.0, 1.0}}, S{P{0.0, 0.5}, P{0.0, 1.0}}});
+  auto expected_merged_flag = make_device_vector<uint8_t>({0, 1});
+
+  find_and_combine_segment(range(segment_offsets.begin(), segment_offsets.end()),
+                           range(segments.begin(), segments.end()),
+                           merged_flag.begin(),
+                           this->stream());
+
+  expect_segment_equivalent(segments, expected_segments);
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(merged_flag, expected_merged_flag);
+}
+
+TYPED_TEST(FindAndCombineSegmentsTest, Touching2)
+{
+  using T       = TypeParam;
+  using index_t = std::size_t;
+  using P       = vec_2d<T>;
+  using S       = segment<T>;
+
+  auto segment_offsets = make_device_vector<index_t>({0, 2});
+  auto segments =
+    make_device_vector<segment<T>>({S{P{0.0, 0.0}, P{0.5, 0.0}}, S{P{0.5, 0.0}, P{1.0, 0.0}}});
+  auto merged_flag = rmm::device_vector<uint8_t>(segments.size());
+
+  auto expected_segments =
+    make_device_vector<segment<T>>({S{P{0.0, 0.0}, P{1.0, 0.0}}, S{P{0.5, 0.0}, P{1.0, 0.0}}});
+  auto expected_merged_flag = make_device_vector<uint8_t>({0, 1});
+
+  find_and_combine_segment(range(segment_offsets.begin(), segment_offsets.end()),
+                           range(segments.begin(), segments.end()),
+                           merged_flag.begin(),
+                           this->stream());
+
+  expect_segment_equivalent(segments, expected_segments);
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(merged_flag, expected_merged_flag);
+}
+
+TYPED_TEST(FindAndCombineSegmentsTest, Touching3)
+{
+  using T       = TypeParam;
+  using index_t = std::size_t;
+  using P       = vec_2d<T>;
+  using S       = segment<T>;
+
+  auto segment_offsets = make_device_vector<index_t>({0, 2});
+  auto segments =
+    make_device_vector<segment<T>>({S{P{0.0, 0.0}, P{0.5, 0.5}}, S{P{0.5, 0.5}, P{1.0, 1.0}}});
+  auto merged_flag = rmm::device_vector<uint8_t>(segments.size());
+
+  auto expected_segments =
+    make_device_vector<segment<T>>({S{P{0.0, 0.0}, P{1.0, 1.0}}, S{P{0.5, 0.5}, P{1.0, 1.0}}});
+  auto expected_merged_flag = make_device_vector<uint8_t>({0, 1});
+
+  find_and_combine_segment(range(segment_offsets.begin(), segment_offsets.end()),
+                           range(segments.begin(), segments.end()),
+                           merged_flag.begin(),
+                           this->stream());
+
+  expect_segment_equivalent(segments, expected_segments);
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(merged_flag, expected_merged_flag);
+}
```

Comment:
Why do the segments merge into a combination in `expected[0]` but then `expected[1] == input[1]`? Does `expected_merged_flag` indicate that `expected[0]` doesn't need to be merged since it is merged, but `expected[1]` can still be merged?